File: backend/app/routers/plans.py
"""REST endpoints for experiment plan CRUD and generation.

Provides plan listing, retrieval, editing, approval, deletion,
and workflow-backed generation via Temporal.

Reference: deep_research router patterns for auth and response envelope.
"""

import logging
import uuid
from datetime import datetime, timezone
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.dependencies import get_current_user, get_db
from app.models.deep_research import DeepResearchTask
from app.models.experiment_plan import ExperimentPlan
from app.models.user import User
from app.schemas.common import ApiResponse
from app.schemas.plan import (
    ExperimentPlanResponse,
    ExperimentPlanUpdate,
    PlanGenerationInput,
)
from app.services.temporal_service import start_workflow
from app.workflows.plan_generation import (
    PlanGenerationWorkflow,
    PlanGenerationWorkflowInput,
)

logger = logging.getLogger(__name__)

router = APIRouter()


@router.post(
    "/generate",
    response_model=ApiResponse[dict],
    status_code=status.HTTP_202_ACCEPTED,
)
async def generate_plans(
    body: PlanGenerationInput,
    user: Annotated[User, Depends(get_current_user)],
    session: Annotated[AsyncSession, Depends(get_db)],
) -> ApiResponse[dict]:
    """Start plan generation workflow for a completed research task.

    Validates task ownership and completion status, then starts
    a Temporal workflow for durable plan generation.
    """
    # Validate task exists and belongs to user
    result = await session.execute(
        select(DeepResearchTask).where(DeepResearchTask.id == body.task_id)
    )
    task = result.scalar_one_or_none()

    if task is None:
        raise HTTPException(status_code=404, detail="Research task not found")
    if task.user_id != user.id:
        raise HTTPException(status_code=403, detail="Not authorized")
    if task.status != "completed":
        raise HTTPException(
            status_code=409,
            detail="Research task must be completed before generating plans",
        )

    # Start Temporal workflow
    workflow_id = f"plan-gen-{body.task_id}-{uuid.uuid4().hex[:8]}"

    workflow_input = PlanGenerationWorkflowInput(
        task_id=body.task_id,
        user_id=user.id,
        entry_type=body.entry_type,
        source_paper_id=body.source_paper_id,
        source_gap_index=body.source_gap_index,
        data_strategy=body.data_strategy,
        num_plans=body.num_plans,
    )

    try:
        await start_workflow(
            workflow_class=PlanGenerationWorkflow,
            workflow_id=workflow_id,
            args=workflow_input,
        )
    except Exception as exc:
        logger.error("Failed to start plan generation workflow: %s", exc)
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail="Workflow service unavailable",
        ) from exc

    return ApiResponse(
        success=True,
        data={
            "workflow_id": workflow_id,
            "task_id": body.task_id,
            "num_plans": body.num_plans,
        },
        message="Plan generation started",
    )


@router.get(
    "/",
    response_model=ApiResponse[list[ExperimentPlanResponse]],
)
async def list_plans(
    user: Annotated[User, Depends(get_current_user)],
    session: Annotated[AsyncSession, Depends(get_db)],
    task_id: str | None = Query(default=None),
    plan_status: str | None = Query(default=None, alias="status"),
    skip: int = Query(default=0, ge=0),
    limit: int = Query(default=20, ge=1, le=100),
) -> ApiResponse[list[ExperimentPlanResponse]]:
    """List experiment plans for the current user.

    Supports filtering by task_id and status, with pagination.
    """
    query = (
        select(ExperimentPlan)
        .where(ExperimentPlan.user_id == user.id)
        .order_by(ExperimentPlan.created_at.desc())
    )

    if task_id is not None:
        query = query.where(ExperimentPlan.task_id == task_id)
    if plan_status is not None:
        query = query.where(ExperimentPlan.status == plan_status)

    query = query.offset(skip).limit(limit)

    result = await session.execute(query)
    plans = list(result.scalars().all())

    return ApiResponse(
        success=True,
        data=[ExperimentPlanResponse.model_validate(p) for p in plans],
    )


@router.get(
    "/{plan_id}",
    response_model=ApiResponse[ExperimentPlanResponse],
)
async def get_plan(
    plan_id: str,
    user: Annotated[User, Depends(get_current_user)],
    session: Annotated[AsyncSession, Depends(get_db)],
) -> ApiResponse[ExperimentPlanResponse]:
    """Get a single experiment plan by ID. Requires ownership."""
    result = await session.execute(
        select(ExperimentPlan).where(ExperimentPlan.id == plan_id)
    )
    plan = result.scalar_one_or_none()

    if plan is None:
        raise HTTPException(status_code=404, detail="Plan not found")
    if plan.user_id != user.id:
        raise HTTPException(status_code=403, detail="Not authorized")

    return ApiResponse(
        success=True,
        data=ExperimentPlanResponse.model_validate(plan),
    )


@router.patch(
    "/{plan_id}",
    response_model=ApiResponse[ExperimentPlanResponse],
)
async def update_plan(
    plan_id: str,
    body: ExperimentPlanUpdate,
    user: Annotated[User, Depends(get_current_user)],
    session: Annotated[AsyncSession, Depends(get_db)],
) -> ApiResponse[ExperimentPlanResponse]:
    """Update a draft plan's fields. Only draft plans can be edited."""
    result = await session.execute(
        select(ExperimentPlan).where(ExperimentPlan.id == plan_id)
    )
    plan = result.scalar_one_or_none()

    if plan is None:
        raise HTTPException(status_code=404, detail="Plan not found")
    if plan.user_id != user.id:
        raise HTTPException(status_code=403, detail="Not authorized")
    if plan.status != "draft":
        raise HTTPException(
            status_code=409,
            detail="Only draft plans can be edited",
        )

    # Apply updates (only non-None fields)
    update_data = body.model_dump(exclude_unset=True)
    for field_name, value in update_data.items():
        setattr(plan, field_name, value)

    plan.updated_at = datetime.now(timezone.utc)
    await session.commit()
    await session.refresh(plan)

    return ApiResponse(
        success=True,
        data=ExperimentPlanResponse.model_validate(plan),
        message="Plan updated",
    )


@router.delete(
    "/{plan_id}",
    status_code=status.HTTP_204_NO_CONTENT,
)
async def delete_plan(
    plan_id: str,
    user: Annotated[User, Depends(get_current_user)],
    session: Annotated[AsyncSession, Depends(get_db)],
) -> None:
    """Delete a draft plan. Only draft plans can be deleted."""
    result = await session.execute(
        select(ExperimentPlan).where(ExperimentPlan.id == plan_id)
    )
    plan = result.scalar_one_or_none()

    if plan is None:
        raise HTTPException(status_code=404, detail="Plan not found")
    if plan.user_id != user.id:
        raise HTTPException(status_code=403, detail="Not authorized")
    if plan.status != "draft":
        raise HTTPException(
            status_code=409,
            detail="Only draft plans can be deleted",
        )

    await session.delete(plan)
    await session.commit()


@router.post(
    "/{plan_id}/approve",
    response_model=ApiResponse[ExperimentPlanResponse],
)
async def approve_plan(
    plan_id: str,
    user: Annotated[User, Depends(get_current_user)],
    session: Annotated[AsyncSession, Depends(get_db)],
) -> ApiResponse[ExperimentPlanResponse]:
    """Approve a draft plan, changing status to 'approved'."""
    result = await session.execute(
        select(ExperimentPlan).where(ExperimentPlan.id == plan_id)
    )
    plan = result.scalar_one_or_none()

    if plan is None:
        raise HTTPException(status_code=404, detail="Plan not found")
    if plan.user_id != user.id:
        raise HTTPException(status_code=403, detail="Not authorized")
    if plan.status != "draft":
        raise HTTPException(
            status_code=409,
            detail="Only draft plans can be approved",
        )

    plan.status = "approved"
    plan.updated_at = datetime.now(timezone.utc)
    await session.commit()
    await session.refresh(plan)

    return ApiResponse(
        success=True,
        data=ExperimentPlanResponse.model_validate(plan),
        message="Plan approved",
    )

File: apps/web/src/lib/api/plans.ts
/**
 * API client for experiment plans.
 *
 * Provides typed functions for all plan CRUD operations,
 * plan generation, and approval workflows.
 */

import { apiFetch } from '../api';

// ─── TypeScript Interfaces ──────────────────────────────────────────────

export interface Baseline {
  readonly name: string;
  readonly paper_id?: string;
  readonly metrics?: Record<string, unknown>;
}

export interface Dataset {
  readonly name: string;
  readonly url?: string;
  readonly size?: string;
  readonly license?: string;
}

export interface RoadmapStep {
  readonly step: number;
  readonly description: string;
}

export interface FeasibilityScore {
  readonly compute_requirements: number;
  readonly data_availability: number;
  readonly expected_improvement: number;
  readonly difficulty: number;
  readonly overall: number;
  readonly explanation: string;
}

export interface ExperimentPlan {
  readonly id: string;
  readonly user_id: string;
  readonly task_id: string;
  readonly entry_type: 'direction' | 'paper' | 'gap';
  readonly source_paper_id: string | null;
  readonly source_gap_index: number | null;
  readonly title: string;
  readonly hypothesis: string;
  readonly method_description: string;
  readonly baselines: readonly Baseline[];
  readonly metrics: readonly string[];
  readonly datasets: readonly Dataset[];
  readonly technical_roadmap: readonly RoadmapStep[];
  readonly code_skeleton: string | null;
  readonly feasibility: FeasibilityScore | null;
  readonly data_strategy: 'open_source' | 'own_data' | 'hybrid';
  readonly status: 'draft' | 'approved' | 'executing' | 'completed';
  readonly created_at: string;
  readonly updated_at: string | null;
}

export interface PlanGenerationInput {
  readonly task_id: string;
  readonly entry_type: 'direction' | 'paper' | 'gap';
  readonly source_paper_id?: string;
  readonly source_gap_index?: number;
  readonly data_strategy?: 'open_source' | 'own_data' | 'hybrid';
  readonly num_plans?: number;
}

export interface PlanGenerationResult {
  readonly workflow_id: string;
  readonly task_id: string;
}

export interface PlanUpdate {
  readonly title?: string;
  readonly hypothesis?: string;
  readonly method_description?: string;
  readonly baselines?: readonly Baseline[];
  readonly metrics?: readonly string[];
  readonly datasets?: readonly Dataset[];
  readonly technical_roadmap?: readonly RoadmapStep[];
  readonly code_skeleton?: string;
  readonly data_strategy?: 'open_source' | 'own_data' | 'hybrid';
}

// ─── API Functions ──────────────────────────────────────────────────────

export async function fetchPlans(
  taskId?: string,
  status?: string,
  skip: number = 0,
  limit: number = 20,
): Promise<readonly ExperimentPlan[]> {
  const params = new URLSearchParams();
  if (taskId) params.set('task_id', taskId);
  if (status) params.set('status', status);
  params.set('skip', String(skip));
  params.set('limit', String(limit));

  const query = params.toString();
  const result = await apiFetch<readonly ExperimentPlan[]>(
    `/api/v1/plans/?${query}`,
  );

  if (result.success) {
    return result.data;
  }
  throw new Error(result.error);
}

export async function fetchPlan(planId: string): Promise<ExperimentPlan> {
  const result = await apiFetch<ExperimentPlan>(
    `/api/v1/plans/${planId}`,
  );

  if (result.success) {
    return result.data;
  }
  throw new Error(result.error);
}

export async function generatePlans(
  input: PlanGenerationInput,
): Promise<PlanGenerationResult> {
  const result = await apiFetch<PlanGenerationResult>(
    '/api/v1/plans/generate',
    {
      method: 'POST',
      body: JSON.stringify(input),
    },
  );

  if (result.success) {
    return result.data;
  }
  throw new Error(result.error);
}

export async function updatePlan(
  planId: string,
  updates: PlanUpdate,
): Promise<ExperimentPlan> {
  const result = await apiFetch<ExperimentPlan>(
    `/api/v1/plans/${planId}`,
    {
      method: 'PATCH',
      body: JSON.stringify(updates),
    },
  );

  if (result.success) {
    return result.data;
  }
  throw new Error(result.error);
}

export async function deletePlan(planId: string): Promise<void> {
  const result = await apiFetch<null>(`/api/v1/plans/${planId}`, {
    method: 'DELETE',
  });

  if (!result.success) {
    throw new Error(result.error);
  }
}

export async function approvePlan(
  planId: string,
): Promise<ExperimentPlan> {
  const result = await apiFetch<ExperimentPlan>(
    `/api/v1/plans/${planId}/approve`,
    {
      method: 'POST',
    },
  );

  if (result.success) {
    return result.data;
  }
  throw new Error(result.error);
}
